src/v/pandaproxy/server.cc

```diff
@@ -80,33 +80,28 @@ struct handler_adaptor : ss::httpd::handler_base {
       const ss::sstring&,
       std::unique_ptr<ss::request> req,
       std::unique_ptr<ss::reply> rep) final {
-        return ss::try_with_gate(
-          _pending_requests,
-          [this,
-           req{std::move(req)},
-           rep{std::move(rep)},
-           m = _probe.hist().auto_measure()]() mutable {
-              server::request_t rq{std::move(req), this->_ctx};
-              server::reply_t rp{std::move(rep)};
-              auto req_size = get_request_size(*rq.req);
-
-              return ss::with_semaphore(
-                       _ctx.mem_sem,
-                       req_size,
-                       [this, rq{std::move(rq)}, rp{std::move(rp)}]() mutable {
-                           if (_ctx.as.abort_requested()) {
-                               set_reply_unavailable(*rp.rep);
-                               return ss::make_ready_future<
-                                 std::unique_ptr<ss::reply>>(std::move(rp.rep));
-                           }
-                           return _handler(std::move(rq), std::move(rp))
-                             .then([](server::reply_t rp) {
-                                 set_mime_type(*rp.rep, rp.mime_type);
-                                 return std::move(rp.rep);
-                             });
-                       })
-                .finally([m{std::move(m)}]() {});
-          });
+        auto measure = _probe.auto_measure();
+        auto guard = gate_guard(_pending_requests);
+        server::request_t rq{std::move(req), this->_ctx};
+        server::reply_t rp{std::move(rep)};
+        auto req_size = get_request_size(*rq.req);
+        auto sem_units = co_await ss::get_units(_ctx.mem_sem, req_size);
+        if (_ctx.as.abort_requested()) {
+            set_reply_unavailable(*rp.rep);
+            measure.set_status(rp.rep->_status);
+            co_return std::move(rp.rep);
+        }
+        try {
+            rp = co_await _handler(std::move(rq), std::move(rp));
+        } catch (const std::exception& e) {
+            auto eptr = std::current_exception();
+            auto rep = exception_reply(eptr);
```

Comment:
This does not capture failures at the seastar level. For example, if you inject a throw in [reply.cc:224](https://github.com/redpanda-data/seastar/blob/0fd17f4cb57365c4a340fbb4435fcc903871512a/src/http/reply.cc#L224), the failure is not counted.

A possible quick-fix is to pass a pointer for `http_status_metric::measurement` into exception_reply and capture the status from there.

Reply:
Will work on this in the morning.

Reply:
It could fail writing the body after writing a 4xx status code to the wire and arrived on the other end of the connection and the response body is still streaming. Should a 4xx or 5xx be recorded?

You probably don't want to count the error twice, so you could wrap `rep` in an exception wrapper, and if it fails at some later stage and arrives at exception_replier, unwrap that exception type, decrement the bucket for rep._status_code and add to the 5xx bucket, but I'm not sure very much is gained. If a 4xx code occurs internally and the client disconnects, should a 5xx be recorded? Perhaps nothing should be recorded? Maybe it depends on whether the client intended to disconnect or whether it was an unintended network failure?

It'll never be perfect. The primary goal of the try catch here is to remove known errors at the start of the response from the (internal) latency report.

Reply:
Having said that, the 4xx are probably more malformed requests, and the 5xx are possible likely due to networking issues between say, the REST proxy and Redpanda, but I think that's ok. Also some of the responses have the wrong code (IMHO), so it can't really be perfect without having an entirely separate signalling mechanism than the status code,

Reply:
This was my late night idea. https://github.com/BenPope/redpanda/pull/41